plugins/analysis-icu/src/main/java/org/elasticsearch/index/analysis/IcuTokenizerFactory.java

```diff
@@ -19,28 +19,119 @@
 
 package org.elasticsearch.index.analysis;
 
+import com.ibm.icu.lang.UCharacter;
+import com.ibm.icu.lang.UProperty;
+import com.ibm.icu.lang.UScript;
+import com.ibm.icu.text.BreakIterator;
+import com.ibm.icu.text.RuleBasedBreakIterator;
 import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.icu.segmentation.DefaultICUTokenizerConfig;
 import org.apache.lucene.analysis.icu.segmentation.ICUTokenizer;
+import org.apache.lucene.analysis.icu.segmentation.ICUTokenizerConfig;
+import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.inject.assistedinject.Assisted;
+import org.elasticsearch.common.io.FileSystemUtils;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.settings.IndexSettings;
 
-import java.io.Reader;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
 
 /**
  */
 public class IcuTokenizerFactory extends AbstractTokenizerFactory {
 
+    private final ICUTokenizerConfig config;
+    private static final String RULE_FILES = "rule_files";
+
     @Inject
-    public IcuTokenizerFactory(Index index, @IndexSettings Settings indexSettings, @Assisted String name, @Assisted Settings settings) {
+    public IcuTokenizerFactory(Index index, @IndexSettings Settings indexSettings, Environment env, @Assisted String name, @Assisted Settings settings) {
         super(index, indexSettings, name, settings);
+        config = getIcuConfig(env, settings);
     }
 
     @Override
     public Tokenizer create() {
-        return new ICUTokenizer();
+        if (config == null) {
+            return new ICUTokenizer();
+        }else{
+            return new ICUTokenizer(config);
+        }
+    }
+
+    public ICUTokenizerConfig getIcuConfig(Environment env, Settings settings) {
+        Map<Integer, String> tailored = new HashMap<>();
+
+        try {
+            String[] ruleFiles = settings.getAsArray(RULE_FILES);
+            for (String scriptAndResourcePath : ruleFiles) {
+                int colonPos = scriptAndResourcePath.indexOf(":");
+                if (colonPos == -1 || colonPos == scriptAndResourcePath.length() - 1) {
+                    throw new IllegalArgumentException(RULE_FILES + " should contain comma-separated \"code:rulefile\" pairs");
+                }
+
+                String scriptCode = scriptAndResourcePath.substring(0, colonPos).trim();
+                String resourcePath = scriptAndResourcePath.substring(colonPos + 1).trim();
+                tailored.put(UCharacter.getPropertyValueEnum(UProperty.SCRIPT, scriptCode), resourcePath);
```

Comment:
I did not check this in detail but if `UCharacter.getPropertyValueEnum()` returns values > `UScript.CODE_LIMIT`, then it would break your code that populates the `breakers` array below.